### PR TITLE
Add TS types

### LINKED ISSUE #910: Proposal: bundle type definitions in Pino itself
## References
I found some prior talkings about this: https://github.com/pinojs/pino/issues/344

## Background
I am opening this issue since we have currently a problem in the Fastify codebase: https://github.com/fastify/fastify/pull/2550

## TLDR;
Fastify relies on Pino for logging and therefore also Fastify's typing has to reference Pino's one. The issue is that at the present day they aren't maintained by Pino's team and we are not going to include 3rd party typings in Fastify codebase. 

## Possible solution
If the Pino core team agree I can actively maintain typings inside this repo with the help of some other fastify core team members:
- https://github.com/fastify/fastify/pull/2550#discussion_r487243232
- https://github.com/fastify/fastify/pull/2550#discussion_r487271985

 I am opening this issue because far as I can see, we don't have any other viable solution for the issue https://github.com/fastify/fastify/pull/2550

### PR DESCRIPTION
**Pull request by kibertoad**
fixes #910

Note that there is definitely room for improvements test-wise and some explicit `tsd` assertions definitely would be helpful, but I think everything else can be done incrementally - this provides a usable foundation.

### CODE REVIEW

**Comment by kibertoad** [comment]
@jsumners I would assume that me and @fox1t would need maintainer permissions if we move forward with the "at least 2 dedicated maintainers for TS" plan with 30 days response SLA. Considering that both of us are already members of fastify org, hopefully this shouldn't be a controversial request.

**Comment by kibertoad** [comment]
If nobody minds, I can also add types to https://github.com/pinojs/pino-std-serializers so that there are no third-party types used altogether (other than official Node ones).

File: package.json
Comment (jsumners) [inline package.json:9]: @mcollina and @davidmarkclements I thought we removed this difficult to maintain property from the `package.json`?

Comment (kibertoad) [reply package.json:9]: `fastify` uses `.npmignore` instead. If that is the way, I can use same approach here as well.

Comment (mcollina) [reply package.json:9]: I didn't remember we used this here even. Let's not change it in this PR.

File: pino.d.ts
Comment (mcollina) [inline pino.d.ts:794]: This shipped in sonic-boom.

File: package.json
Comment (jsumners) [reply package.json:9]: Agreed. Just asking because it got added in `pino-std-serializers` with this types stuff.

File: pino.d.ts
Comment (kibertoad) [reply pino.d.ts]: Changed to use it.

File: package.json
Comment (kibertoad) [reply package.json:9]: @jsumners pino-std-serializer currently bundles all the garbage such as travis config, which felt not ideal. I can remove it from there if you feel strongly about it.

Comment (jsumners) [reply package.json:9]: Looks like we actually settled on `files` 🤦‍♂️ 

https://github.com/pinojs/pino/issues/136

**Review by jstewmon — COMMENTED** [review]
Several type and interface definitions are not exported, which can be inconvenient for users. Interfaces need to be exported to support module augmentation, and, in general, exporting all type and interface definitions offers convenience to users because they can more easily use them to declare types in their own code that interoperates with the package.

For example, `LevelChangeEventListener` is useful for package users to declare the type of a function which will be passed as a listener.

**Comment by 72636c** [comment]
Thanks for this work! Unfortunately the default-exported types in `sonic-boom@1.2.0` are now incompatible with `@types/pino`. Would it be worth patching `@types/pino` and `@types/sonic-boom` to switch to the default export while this is in flux?

```
node_modules/@types/pino/index.d.ts(194,122): error TS2709: Cannot use namespace 'SonicBoom' as a type.
node_modules/@types/pino/index.d.ts(202,57): error TS2709: Cannot use namespace 'SonicBoom' as a type.
```

**Comment by mcollina** [comment]
From the amount of feedbacks we are getting from _just_ changing sonic-boom types, I think we must do this as a semver-major change.

**Comment by mcollina** [comment]
I'll revert in sonic-boom as well. Let's treat them all as semver-major.

File: pino.d.ts
Comment (kibertoad) [reply pino.d.ts]: Done

**Comment by kibertoad** [comment]
> 
> 
> Several type and interface definitions are not exported, which can be inconvenient for users. Interfaces need to be exported to support module augmentation, and, in general, exporting all type and interface definitions offers convenience to users because they can more easily use them to declare types in their own code that interoperates with the package.
> 
> For example, `LevelChangeEventListener` is useful for package users to declare the type of a function which will be passed as a listener.

Please take a look at the latest version, I exported the namespace, now these definition can be reached from within the namespace. Alternatively I think we can try to get rid of namespace altogether and do it more similarly to how fastify does it - but that would be a way more intrusive and breaking change.

File: pino.js
Comment (fox1t) [inline pino.js:231]: Why `P` and not `pino`?

File: pino.d.ts
Comment (fox1t) [inline pino.d.ts:22]: It would be much better to use Fastify approach and have the "infamous tripet" exported in JS and here a default + named export without using namespace at all. It is the only known way we have to support all of the tsconfigurations, faux modules, bundlers and so on ( :( )

**Review by fox1t — CHANGES_REQUESTED** [review]

**Comment by fox1t** [comment]
@kibertoad I think we should go with the fastify apporach even if it takes a longer to be merged. It is surely more stable and better done.

**Comment by kibertoad** [comment]
@fox1t You are right. Since we are going for semver major either way, might as well use the opportunity to refine the API.

File: pino.js
Comment (kibertoad) [reply pino.js:231]: to preserve naming of original `@types/pino`, but you are right, we don't really need to do that.

**Comment by mcollina** [comment]
After the current incident, it seems that a significant amount of pino users are served by the types in `@types/pino`. Moving them to these will break them badly. I'll need to let this bounce a bit in my head before making a call. By adding types to pino, we will be adding _significant more issue traffic_ that this project currently gets. Logging should be "boring" and work well.

**Comment by jsumners** [comment]
> By adding types to pino, we will be adding _significant more issue traffic_ that this project currently gets. Logging should be "boring" and work well.

Agreed. And increasing the issue load around here is not helpful.

**Comment by fox1t** [comment]
Yea, there are plenty of devs using pino types. I agree we should not break them. If we don't see any benefit from porting them in this package and allign them to fastify ones (triplet and so on) there is no point in breaking anything.

**Comment by kibertoad** [comment]
@mcollina There are multiple directions where we can go, though. If we want to avoid breaking changes, we can, there is nothing stopping us from adopting existing DefinitelyTyped typings verbatim so that there is zero chance to break anything for anyone. And we'll still get a better workflow and more agility for how we address future challenges - when there will be pino 7 on the horizon, we could seize the opportunity and implement breaking changes _then_, cleaning up the types and making them more robust.
And obviously there are two sides to "lots of people use `@types/pino`" as well - it means that any improvements that we make will have greater impact as well, and I am very confident that whatever challenges people may encounter with types, we would be able to address them faster and better than what is currently the case with DefinitelyTyped ad-hoc maintainership.

**Comment by davidmarkclements** [comment]
is anyone still wanting to maintain types in Pino? If not I'll close this in a week or so.

**Comment by kibertoad** [comment]
@davidmarkclements Yes. Is next semver major coming up? Can we get back to getting this merged?

**Comment by davidmarkclements** [comment]
@kibertoad major depends on progress in https://github.com/pinojs/pinole

**Comment by kibertoad** [comment]
@mcollina Can we start prepping by releasing semver major for all the downstream deps on which types pino rely like sonic-boom?

**Comment by mcollina** [comment]
> Can we start prepping by releasing semver major for all the downstream deps on which types pino rely like sonic-boom?

I would wait until pinole is finished.

**Comment by mcollina** [comment]
@kibertoad are you still up to support TS in pino in v7? We are possibly approaching a release after https://github.com/pinojs/pino/pull/1003 lands.

**Comment by kibertoad** [comment]
@mcollina Yup. Is it a good time to start going over downstream deps again, releasing semver majors with types?

**Comment by mcollina** [comment]
Yes, that's the idea. This PR should target the `next` branch.

**Comment by kibertoad** [comment]
@mcollina Some changes leaked from `master` that are not yet in `next` into this PR. Can I create a PR syncing `next` with `master` to simplify things?

File: package.json
Comment (kibertoad) [inline package.json:102]: this will be replaced with released version after it is available

**Comment by mcollina** [comment]
> @mcollina Some changes leaked from `master` that are not yet in `next` into this PR. Can I create a PR syncing `next` with `master` to simplify things?

This should be fixed now.

Note that there is also https://github.com/pinojs/pino/pull/1003 that would need typings.

**Comment by kibertoad** [comment]
@mcollina Sure. Would you like these types added in this PR, or should I wait until #1003 lands and create separate PR then?

File: test/types/pino.test-d.ts
Comment (mcollina) [inline test/types/pino.test-d.ts:32]: Does this use tabs for indentation? Maybe can we have this conform to the rest?

**Comment by kibertoad** [comment]
Will update the PR today

File: test/types/pino.test-d.ts
Comment (kibertoad) [reply test/types/pino.test-d.ts:32]: It already uses spaces.

File: pino.d.ts
Comment (kibertoad) [reply pino.d.ts]: Removed everything referring to extreme mode

File: package.json
Comment (kibertoad) [reply package.json]: Fixed

**Comment by kibertoad** [comment]
@mcollina Ready for re-review!

File: package.json
Comment (mcollina) [inline package.json:25]: can you move the type tests to the end?

**Review by mcollina — CHANGES_REQUESTED** [review]
Multistream is missing

File: package.json
Comment (kibertoad) [reply package.json]: done

**Comment by kibertoad** [comment]
@mcollina Typed multistream to the best of my ability based on API docs, could you please check if typing makes sense?

File: pino.d.ts
Comment (mcollina) [inline pino.d.ts:2]: I think this header should be changed/updated.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by mcollina** [comment]
@kibertoad I've invited you to this org to make sure you can help maintain these going forward.

**Pull request merged**
